File: api/src/main/java/com/example/api/Document/Document.java
package com.example.api.Document;

import com.example.api.User.User;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;

import java.io.Serializable;

@Setter
@Getter
@Entity
@Table(name = "documents")
public class Document implements Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull(message = "Người hướng dẫn là bắt buộc")
    @Column(length = 255, nullable = false)
    private String advisor;

    @NotNull(message = "Tác giả là bắt buộc")
    @Column(length = 255, nullable = false)
    private String author;

    @NotNull(message = "Năm tốt nghiệp là bắt buộc")
    @Column(nullable = false)
    private int graduationYear;

    @NotNull(message = "Dạng tài liệu là bắt buộc")
    @Column(length = 50, nullable = false)
    private String type;

    @NotNull(message = "Chuyên ngành là bắt buộc")
    @Column(length = 100, nullable = false)
    private String specialization;

    @NotNull(message = "Tóm tắt là bắt buộc")
    @Column(columnDefinition = "TEXT", nullable = false)
    private String abstractText;

    @NotNull(message = "Ngôn ngữ là bắt buộc")
    @Column(length = 50, nullable = false)
    private String language;

    @NotNull(message = "Nhà xuất bản là bắt buộc")
    @Column(length = 255, nullable = false)
    private String publisher;

    @NotNull(message = "Nhan đề là bắt buộc")
    @Column(length = 255, nullable = false)
    private String title;

    @NotNull(message = "Từ khóa là bắt buộc")
    @Column(columnDefinition = "TEXT", nullable = false)
    private String keywords;

    @Column(length = 255, nullable = false)
    private String pdfFileName;

    @Column(length = 255, nullable = false)
    private String pdfUrl;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "userId", nullable = false)
    @JsonIgnore // Tránh vòng lặp khi tuần tự hóa JSON
    private User user;

    // Constructors
    public Document() {
    }

    public Document(String advisor, String author, int graduationYear, String type,
                    String specialization, String abstractText, String language,
                    String publisher, String title, String keywords, String pdfFileName,String pdfUrl) {
        this.advisor = advisor;
        this.author = author;
        this.graduationYear = graduationYear;
        this.type = type;
        this.specialization = specialization;
        this.abstractText = abstractText;
        this.language = language;
        this.publisher = publisher;
        this.title = title;
        this.keywords = keywords;
        this.pdfFileName = pdfFileName;
        this.pdfUrl = pdfUrl;
    }

}

File: lib/models/document/document.dart
class Document {
  final int? id;
  late final String advisor;
  late final String author;
  late final int graduationYear;
  late final String type;
  late final String specialization;
  late final String abstractText;
  late final String language;
  late final String publisher;
  late final String title;
  late final String keywords;
  late final String pdfFileName;
  late final String pdfUrl;

  Document({
    this.id,
    required this.advisor,
    required this.author,
    required this.graduationYear,
    required this.type,
    required this.specialization,
    required this.abstractText,
    required this.language,
    required this.publisher,
    required this.title,
    required this.keywords,
    required this.pdfFileName,
    required this.pdfUrl,
  });

  factory Document.fromMap(Map<dynamic, dynamic> documentMap) {
    return Document(
      id: documentMap['id'] as int?,
      advisor: documentMap['advisor'] as String,
      author: documentMap['author'] as String,
      graduationYear: documentMap['graduationYear'] as int,
      type: documentMap['type'] as String,
      specialization: documentMap['specialization'] as String,
      abstractText: documentMap['abstractText'] as String,
      language: documentMap['language'] as String,
      publisher: documentMap['publisher'] as String,
      title: documentMap['title'] as String,
      keywords: documentMap['keywords'] as String,
      pdfFileName: documentMap['pdfFileName'] as String,
      pdfUrl: documentMap['pdfUrl'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'advisor': advisor,
      'author': author,
      'graduationYear': graduationYear,
      'type': type,
      'specialization': specialization,
      'abstractText': abstractText,
      'language': language,
      'publisher': publisher,
      'title': title,
      'keywords': keywords,
      'pdfFileName': pdfFileName,
      'pdfUrl': pdfUrl,
    };
  }
}
